### DISCUSSION TITLE
Support for CockroachDB

### DISCUSSION THREAD

**Discussion by sandstrom** [bensheldon/good_job#292] [General]
CockroachDB -- a new database, aspiring to be wire-compatible with PostgreSQL -- has an ActiveRecord adapter that works fairly well (https://github.com/cockroachdb/activerecord-cockroachdb-adapter).

However, they are currently lacking support for [Advisory Locks](https://github.com/cockroachdb/cockroach/issues/13546) and [LISTEN/NOTIFY](https://github.com/cockroachdb/cockroach/issues/41522#issuecomment-874071874) which are two requirements for Good Job.

Just wanted to check, are there any other "special" features in PostgreSQL, that Good Job depends on?

cc @bensheldon

**Comment by bensheldon** [comment]
Off the top of my head, GoodJob uses these features from Postgres:

- Advisory Locks
- LISTEN/NOTIFY
- CTEs (Common Table Expressions)
- Introspection and management features (e.g. `pg_locks`, `pg_stat_activity`, `pg_backend_pid()`, `pg_terminate_backend()`)
- `bigint` column-type and `md5()`

**Reply by sandstrom** [reply]
Thanks!

CTEs, `bigint` and `md5` should exist. 

Don't know about `pg_locks`, `pg_stat_activity`, `pg_backend_pid()`, `pg_terminate_backend()`.

No need to do anything about Cockroach until they have Advisory locks and LISTEN/NOTIFY. Mainly wanted to start a discussion thread, to see what interest there is and to track what's missing from Cockroach currently.

**Comment by jonahgeorge** [comment]
On this topic, have you ever thought about expanding GoodJob to support pluggable backends?

I've always considered GoodJob to be the Rails analogy to [Oban](https://github.com/sorentwo/oban), which recently added support for [Sqlite3](https://github.com/sorentwo/oban/releases/tag/v2.14.0) and [has MySQL on its radar](https://github.com/sorentwo/oban/issues/836). 

For context, there have been a few scenarios where I would have really liked to be able to run GoodJob in a non-Postgres (but still ActiveRecord/Relational DB) environment, accepting that job execution performance would likely suffer due to it requiring a polling approach rather than LISTEN/NOTIFY & advisory locks. 

I fully recognize this is a massive undertaking, but just curious to hear whether this is something you'd even consider for outside contribution.

**Reply by bensheldon** [reply]
👋 Thanks for asking the question. Here's where my head is at today:

I'd like to keep GoodJob Postgres-only. But I'd be overjoyed to help someone build their own Active Job backend ✨

Here's why: 80% of my effort for GoodJob goes into working with Postgres. I think there's still a lot of interesting runway to improve Postgres-compatibility and performance which makes me very reluctant to add another database. And lots of features (like Concurrency Control, Cron, or the Dashboard) that are very much in the category of "huh, I wonder if that's possible?" and the drag of having to implement every expeditionary feature multiple times fills me with dread. 

On the bright side, my learnings from GoodJob is this: building a multithreaded Active Job backend is not a difficult problem. Concurrent Ruby has all the primitives, Rails Engines do a lot of heavy lifting. Executing jobs is not the bottleneck; efficiently fetching and safely locking job data from the database is the hard part.

GoodJob's core job execution objects are: `Poller`, `Scheduler` and the `JobPerformer` interface. They're not very complicated until you look behind the JobPerformer at the Active Record model and anything non-trivial is very Postgres-specific. But, if you copy-paste those 3 objects (which are wrappers around Concurrent Ruby) into a new engine and implement the JobPerformer interface... you've got yourself an Active Job backend (albeit no other bells or whistles). 

(I'm oversimplifying it a smidge, but not really!)

I suggest making your own backend because I'd really, really like to see more people working with Concurrent Ruby. It's a lovely library. You'll probably teach me something working through the problems yourself. And I'd love to help... just not in GoodJob right now 💖

**Reply by jonahgeorge** [reply]
Just dumping some notes here:

<details>
<summary>CockroachDB</summary>

- Advisory Locks ❌
  - https://github.com/cockroachdb/cockroach/issues/13546
  - https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20171024_select_for_update.md
- LISTEN/NOTIFY ❌
  - Fallback to inefficient poller?
  - https://github.com/cockroachdb/cockroach/issues/41522
- CTEs (Common Table Expressions) ✅
- Introspection and management features ❌
- bigint column-type and md5() ✅
  - [`bigint`](https://www.cockroachlabs.com/docs/stable/int.html)
  - [`md5`](https://www.cockroachlabs.com/docs/stable/functions-and-operators.html#string-and-byte-functions)
</details>

<details>
<summary>MySQL</summary>

- Advisory Locks ✅
  - [`get_lock`](https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_get-lock)
  - [`release_lock`](https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_release-lock)
  - [`is_free_lock`](https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_is-free-lock)
  - [`SELECT FOR UPDATE SKIP LOCKED`](https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html)
- LISTEN/NOTIFY ❌
  - Fallback to inefficient poller?
- CTEs (Common Table Expressions) ✅
- Introspection and management features ⚠️
  - [`performance_schema.metadata_locks`](https://dev.mysql.com/doc/refman/5.7/en/performance-schema-metadata-locks-table.html)
  - [`show full processlist`](https://dev.mysql.com/doc/refman/8.0/en/show-processlist.html)
  - [`connection_id()`](https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_connection-id)
  - [`kill`](https://dev.mysql.com/doc/refman/8.0/en/kill.html)
- bigint column-type and md5() ✅
  - [`md5()`](https://dev.mysql.com/doc/refman/8.0/en/encryption-functions.html#function_md5)
  - [`bigint`](https://dev.mysql.com/doc/refman/8.0/en/integer-types.html)
</details>

<details>
<summary>MSSQL</summary>

- Advisory Locks ✅
  - [`sp_getapplock`](https://learn.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-getapplock-transact-sql?view=sql-server-ver16)
  - [`sp_releaseapplock`](https://learn.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-releaseapplock-transact-sql?view=sql-server-ver16)
  - [`applock_test`](https://learn.microsoft.com/en-us/sql/t-sql/functions/applock-test-transact-sql?source=recommendations&view=sql-server-ver16)
  - [`applock_mode`](https://learn.microsoft.com/en-us/sql/t-sql/functions/applock-mode-transact-sql?source=recommendations&view=sql-server-ver16)
  - [Pessimistic Locking](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/issues/368)
    - [`readpast`](https://learn.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql-table?view=sql-server-ver15#readpast)
    - [`updlock`](https://learn.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql-table?view=sql-server-ver15#updlock)
- LISTEN/NOTIFY ❌
  - Fallback to inefficient poller?
  - Use [Service Broker](https://learn.microsoft.com/en-us/sql/database-engine/configure-windows/sql-server-service-broker?view=sql-server-ver16)?
- CTEs (Common Table Expressions) ✅
- Introspection and management features ⚠️
  - [`sys.dm_tran_locks`](https://learn.microsoft.com/en-us/sql/relational-databases/system-dynamic-management-views/sys-dm-tran-locks-transact-sql?redirectedfrom=MSDN&view=sql-server-ver16)
  - [`sp_who`](https://learn.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-who-transact-sql?view=sql-server-ver16)
  - [`kill`](https://learn.microsoft.com/en-us/sql/t-sql/language-elements/kill-transact-sql?view=sql-server-ver16)
  - [`@@spid`](https://learn.microsoft.com/en-us/sql/t-sql/functions/spid-transact-sql?view=sql-server-ver16)
- bigint column-type and md5() ✅
  - [`hashbytes`](https://learn.microsoft.com/en-us/sql/t-sql/functions/hashbytes-transact-sql?view=sql-server-ver16)
  - [`bigint`](https://learn.microsoft.com/en-us/sql/t-sql/data-types/int-bigint-smallint-and-tinyint-transact-sql?view=sql-server-ver16)
</details>

<details>
<summary>Sqlite</summary>

- Advisory Locks ❌
  - https://www.sqlite.org/lockingv3.html
- LISTEN/NOTIFY ❌
- CTEs (Common Table Expressions) ✅
- Introspection and management features ❌
- bigint column-type and md5() ❌
</details>

Misc Links:
- https://github.com/ClosureTree/with_advisory_lock
- [Oban Sqlite engine](https://github.com/sorentwo/oban/blob/main/lib/oban/engines/lite.ex#L87-L116)
- https://www.sqlines.com/oracle-to-mysql/skip_locked
- https://www.postgresql.org/docs/current/sql-select.html#SQL-FOR-UPDATE-SHARE
- [ActiveRecord Pessimistic Locking](https://api.rubyonrails.org/classes/ActiveRecord/Locking/Pessimistic.html)

**Comment by brandonzylstra** [comment]
@sandstrom would it work to run all your jobs in Postgres and keep other data in CockroachDB?  Rails makes it fairly easy to have multiple databases per app, and if necessary Postgres' could have views that read data from CockroachDB (probably using Foreign Data Wrappers—though I don't remember if FDWs are needed to read from another database that "looks like" another PG db), although I don't know why that would actually ever be necessary.

I'd love to know if you get that worked out, since I'm working on an app that is starting on Postgres but which may switch to CockroachDB if it resolves certain scaling issues that I expect might arise—and I'd prefer to keep GoodJob if possible.

@bensheldon I'd also love to hear your thoughts on whether this sounds feasible

**Reply by bensheldon** [reply]
Yep, this sounds similar to the setup for PgBouncer, which builds on GoodJob's compatibility with multiple Active Record database connections (and cross-database data is done via Active Job serializers and GlobalID which in my experience are pretty seamless): 

https://github.com/bensheldon/good_job#pgbouncer-compatibility

You lose the small niceness of transactional integrity when enqueuing going to multiple databases, but otherwise, I don't think you'd notice the difference 👍🏻

**Reply by brandonzylstra** [reply]
Thanks for the confirmation!  It's always nice to have *one less thing to worry about* when looking toward the future 😄

**Reply by sandstrom** [reply]
Ben already answered, but basically yes, that would work. I would also guess that CockroachDB will eventually support the missing pieces.

As Ben mentions, you'd loose transactional integrity when enqueuing jobs though.

If you haven't already, you can always add your 👍🏻-vote in the two CDB-issues referenced above:

- https://github.com/cockroachdb/cockroach/issues/13546
- https://github.com/cockroachdb/cockroach/issues/41522
